### ISSUE TITLE
Roc panics in Json builtin with `assertion failed: value == 0 || value == 1`

### ISSUE THREAD

**Issue by lukewilliamboswell** [roc-lang/roc#5186]
@ayazhafiz can you please help me identify what this bug is in this [gist](https://gist.github.com/lukewilliamboswell/652901f4c5a03f10addf8ee8a263fdaf)? The below test is where you can reproduce the bug. Commenting this test out and everything else passes as expected. From what I can gather there is something that determines which `decoder` to use based on the types the compiler understands (from the `actual` annotation in this case) for the result. However, I think that logic may not be working correctly after playing with this for a bit. If you add additional numbers into the list, it will parse them all expect the last one. I see no reason why the number decoders shouldn't work, having gone through the `decodeList` many times, and the other tests for numbers. Maybe it is the interaction between numbers and lists specifically, though strings and lists work ok.

```
# Test json array decoding into integers
expect
    input = Str.toUtf8 "[1]"
    actual : Result (List U16) _
    actual = Decode.fromBytes input fromUtf8
    expected = Ok [1]

    actual == expected
```
```sh
$ cargo run -- test crates/compiler/builtins/roc/Json.roc 
   (I removed compiling lines here ...)
    Finished dev [unoptimized + debuginfo] target(s) in 15.99s
     Running `target/debug/roc test crates/compiler/builtins/roc/Json.roc`
thread 'main' panicked at 'assertion failed: value == 0 || value == 1', crates/repl_expect/src/app.rs:25:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
```
$ roc test crates/compiler/builtins/roc/Json.roc 
── EXPECT FAILED ─────────────────────── crates/compiler/builtins/roc/Json.roc ─

This expectation failed:

610│>  # Test json array decoding into integers
611│>  expect
612│>      input = Str.toUtf8 "[1]"
613│>      actual : Result (List U16) _
614│>      actual = Decode.fromBytes input fromUtf8
615│>      expected = Ok [1]
616│>
617│>      actual == expected

When it failed, these variables had these values:

input : List U8
input = [91, 49, 93]

actual : Result (List (Int Unsigned16)) [Leftover (List U8), TooShort]
actual = Err (Leftover [49, 93])

expected : [Err [Leftover (List U8), TooShort],
Ok (List (Int Unsigned16))]
expected = Ok [1]



1 failed and 49 passed in 855 ms.
```

**Comment by lukewilliamboswell** [comment]
Ah, looks to be related to #5180

**Comment by lukewilliamboswell** [comment]
From further testing I am reasonably confident the bug is somewhere inside my `takeJsonNumber` function.

**Comment by lukewilliamboswell** [comment]
I managed to resolve this issue, the problem was from the `numberHelp` function, specifically when using `isValidEnd` in the guard and not accounting for the possibility of having a `]` character as a possible byte at the end of a number. 

I cannot say why this threw the assertion, but it was a common bug I kept running into using this pattern, specifically a `when` statement with `state` and `byte` to walk through each byte and update the state. 

This assertion seems to be thrown when there is some possible valid state that isn't correctly identified.
